### DISCUSSION TITLE
Need to include option if explicit edge dataframe already exists, to rewrite/merge

### DISCUSSION THREAD

**Discussion by silkspace** [graphistry/pygraphistry#320] [Ideas]
https://github.com/graphistry/pygraphistry/blob/0f610e90b00232d92e80ca580a1d8e08d0b82ee9/graphistry/feature_utils.py#L1322

**Comment by lmeyerov** [comment]
Some scenarios for a hypothetical `edge_merge` policy, trying to reuse standard relational algebra / sql concepts:

`edge_merge` policies:
* `concat`: fully new edges, preserving existing unaltered. add tag `umap` <-- **default**
* `replace`: fully drop old if any, and only put in umap neighbor edges
* `inner`: only label existing edges, and remove unmatched ones. When multiedges, drop excess.
* `left`: preserve existing edges, and enrich when there's a match. When multiedges, update just 1.
* `upsert`: update existing edges, and preserve umatched. Add new as needed. When multiedges, update just 1.
* `skip`: Preserve initial edges; only enrich nodes w/ the embedding 


append new edges:

```python
(g
  .nodes(genes_df, 'id')
  .edges(known_pathways_df, 'src', 'dst')
   .umap(edge_merge='concat', type_tag='umap', col_prefix='umap_')
   .scale(1.0) # should be responsive
).plot()
```

reuse edges by sticking with either just updating existing (what happens in multiedges, just pick first?) , and optionally adding missing

```python
(g
  .nodes(entities_df, 'id')
  .edges(known_correlations_df, 'src', 'dst')
  .umap(scale=1.0, edge_merge='inner' | 'upsert', type_tag='umap', out_cols={'weight': 'score'}) 
```

use UMAP for 'clustering on some node columns' but preserve existing edges, or at most augmenting them

```python
(g
  .nodes(entities_df, 'id')
  .edges(some_relns_df, 'src', 'dst')
  .umap(use_cols=['feat1', 'feat2'], use_edges=True, scale=1.0, edge_merge='skip' | 'inner')
)
```

use UMAP for community detection, without repositioning

```python
(g
  .nodes(entities_df, 'id')
  .edges(some_relns_df, 'src', 'dst')
  .umap(use_edges=True, scale=1.0, edge_merge='skip', position=False)
  .dbscan(use_cols=['x', 'y'], out_cols={'label': 'community', 'weight': 'community_distance'})
)
```
